This PR introduces a fix for the bug of changing the secret after the certificate is created and valid

arielsepton opened the pull request:
### Changes Made
* Added logic to handle the case where the secretName in the `Certificate` resource was changed when the certificate is valid.
* Updated the SetupWithManager function to watch for `Secret` delete events that have an owner reference of type `Certificate`.

### Testing
* Ran the existing test cases.
* Added a new test case for the bug fix.

### Issue Link
Fixes #10

dependabot[bot] committed:
Bump docker/login-action

Bumps [docker/login-action](https://github.com/docker/login-action) from 5f4866a30a54f16a52d2ecb4a3898e9e424939cf to db779458846adfddbc884a8e8b86376e277d24b2.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/5f4866a30a54f16a52d2ecb4a3898e9e424939cf...db779458846adfddbc884a8e8b86376e277d24b2)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump github.com/go-logr/logr from 1.4.1 to 1.4.2

Bumps [github.com/go-logr/logr](https://github.com/go-logr/logr) from 1.4.1 to 1.4.2.
- [Release notes](https://github.com/go-logr/logr/releases)
- [Changelog](https://github.com/go-logr/logr/blob/master/CHANGELOG.md)
- [Commits](https://github.com/go-logr/logr/compare/v1.4.1...v1.4.2)

---
updated-dependencies:
- dependency-name: github.com/go-logr/logr
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump github.com/crossplane/crossplane-runtime from 1.15.1 to 1.16.0

Bumps [github.com/crossplane/crossplane-runtime](https://github.com/crossplane/crossplane-runtime) from 1.15.1 to 1.16.0.
- [Release notes](https://github.com/crossplane/crossplane-runtime/releases)
- [Changelog](https://github.com/crossplane/crossplane-runtime/blob/master/RELEASE.md)
- [Commits](https://github.com/crossplane/crossplane-runtime/compare/v1.15.1...v1.16.0)

---
updated-dependencies:
- dependency-name: github.com/crossplane/crossplane-runtime
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump docker/login-action

Bumps [docker/login-action](https://github.com/docker/login-action) from db779458846adfddbc884a8e8b86376e277d24b2 to 0d4c9c5ea7693da7b068278f7b52bda2a190a446.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/db779458846adfddbc884a8e8b86376e277d24b2...0d4c9c5ea7693da7b068278f7b52bda2a190a446)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

arielsepton committed:
handle case of secret updates after certificate is valid
Closes #10. This PR introduces a fix for the bug of changing the secret after the certificate is created and valid.

fix CR comments

Signed-off-by: arielsepton <arielsepton1@gmail.com>

mzeevi commented on go.mod:
```diff
@@ -12,7 +12,7 @@ require (
 	go.uber.org/zap v1.27.0
 	k8s.io/api v0.29.4
-	k8s.io/apimachinery v0.29.4
+	k8s.io/apimachinery v0.30.1
```
Can we keep it at `v0.29.4`? This may cause a dependency problem in depending projects

mzeevi commented on internal/controller/certificate_controller.go:
```diff
@@ -79,6 +83,12 @@ type CertificateReconciler struct {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.Certificate{}).
+		Owns(&corev1.Secret{}).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return true
+			},
```
I could be wrong but I'm not sure this actually does any filtering. I _think_ that if you want to filter out `CREATE` and `UPDATE` events then you need to return a false on their predicate

mzeevi commented on internal/controller/certificate_controller.go:
```diff
@@ -201,6 +215,27 @@ func isCertificateValid(certificate *v1alpha1.Certificate, certificateConfig *v1
+// It returns true if the reconciliation process should stop because the secret is up to date or an error occurred.
+// It returns false if the reconciliation should continue.
+func (r *CertificateReconciler) isSecretUpToDate(ctx context.Context, certificate *v1alpha1.Certificate, namespace string) (stopReconcile bool, err error) {
+	if certificate.Status.SecretName != certificate.Spec.SecretName {
+		// If the current secret name doesn't match the desired secret name, continue reconciliation.
+		return false, nil
+	}
+
+	if _, err := common.GetSecret(r.Client, ctx, certificate.Status.SecretName, namespace); err != nil {
+		if errors.IsNotFound(err) {
+			// If the secret was deleted, continue reconciliation
+			return false, nil
+		}
+		return true, err
+	}
```
Break it up to 2 functions: `isSecretNameChanged` and `isSecretDeleted`

mzeevi commented on internal/controller/certificate_controller.go:
```diff
@@ -201,6 +215,27 @@ func isCertificateValid(certificate *v1alpha1.Certificate, certificateConfig *v1
+		return true, err
+	}
+
+	// Secret is up to date, stop reconciliation
```
This is explained from the godoc above the function so remove these comments

mzeevi reviewed (changes requested):
Added some small comments. 

Also please amend the commit message. You don't need to write the full name of the issue in the commit message. You can just write "closes #<issue_number>" and it would automatically link it and close the issue

mzeevi committed:
change rsa block cost to rsa private key
This PR fixes a small bug in the parsing of the certificates. Now the
header of the private key reads RSA PRIVATE KEY which is correct.

Signed-off-by: mzeevi <meytar80@gmail.com>

arielsepton commented:
Dup of #14
